test/integration/fake_upstream.cc

```diff
@@ -413,6 +414,43 @@ void FakeHttpConnection::onGoAway(Http::GoAwayErrorCode code) {
   ENVOY_LOG(info, "FakeHttpConnection receives GOAWAY: ", code);
 }
 
+void FakeHttpConnection::encodeGoAway() {
+  ASSERT(type_ == Type::HTTP2);
+
+  shared_connection_.connection().dispatcher().post([this]() { codec_->goAway(); });
+}
+
+void FakeHttpConnection::encodeProtocolError() {
+  ASSERT(type_ == Type::HTTP2);
+
+  Http::Http2::ServerConnectionImpl* codec =
+      dynamic_cast<Http::Http2::ServerConnectionImpl*>(codec_.get());
+  if (codec != nullptr) {
+    shared_connection_.connection().dispatcher().post([this, codec]() {
+      expect_protocol_error_ = true;
+      codec->protocolErrorForTest();
+      // We'd ideally set expect_protocol_error_ back to false after this, but
+      // it's difficult to sequence that correctly since it needs to be true
+      // until the PROTOCOL_ERROR goaway frame is flushed.
+    });
+    return;
```

Comment:
del

Reply:
If I remove this I need an `else` for the rest of the method, or not assert that one or the other `dynamic_cast` succeeds. Another option would be to make sure we never use the legacy codec? I don't have context on that migration, it seems like only in the `compile_time_options` test is the legacy codec used